### PR TITLE
Clarify container external connectivity in multi-network scenario

### PR DESCRIPTION
**Pull request by aboch**
It is not yet explained in the docs.
Isuse raised in #21741 

Signed-off-by: Alessandro Boch aboch@docker.com

### CODE REVIEW

File: docs/userguide/networking/dockernetworks.md [hunk 1]
```diff
@@ -293,7 +293,9 @@ specifications.
 You can create multiple networks. You can add containers to more than one
 network. Containers can only communicate within networks but not across
 networks. A container attached to two networks can communicate with member
-containers in either network.
+containers in either network. When a container is connected to multiple
+networks, its external connectivity is provided via the first in
```
Comment (thaJeztah) [inline docs/userguide/networking/dockernetworks.md]: Perhaps;

```markdown
...its external connectivity is provided via the first
non-internal network, in lexicographical order.
```

(or "alphabetical order")?

**Comment by darkermatter** [comment]
Please clarify...

1) Lexicographical order as in the Oxford English Dictionary
2) Lexicographical order as in ASCII/UTF
3) Lexicographical order as in an OrderedDict

a) Lexicographical order in the networks section
b) Lexicographical order under each service

Thanks.

**Comment by aboch** [comment]
@darkermatter 
The ordering is achieved by applying the golang comparison operator `<` between the network names strings. As per the go docs, it is a byte-wise lexical ordering. My knowledge about the ordering ends here ;-)
Endpoint name (if this is what you were referring to with "service") is not taken into account.

File: docs/userguide/networking/dockernetworks.md (hunk 1 above)
Comment (aboch) [inline docs/userguide/networking/dockernetworks.md]: Thanks.
Rephrased as adviced, but preferred to be consistent with golang docs, so used "lexical".
Let me know.

**Comment by thaJeztah** [comment]
LGTM

**Comment by vdemeester** [comment]
LGTM 🐶

**Comment by darkermatter** [comment]
Hi @aboch,

Allow me to clarify my request for clarification :-S

In docker-compose.yml v2 files there are networks and services (among others) sections. I define the networks in the networks section and use them in the services section under each service I define.

```
networks:
  network1: ...
  network2: ...

services:
  service1:
    networks:
      - network1
  service2:
    networks:
      - network2
      - network1
```

Where is the order for the container/service derived from?

Cheers.

**Comment by thaJeztah** [comment]
@darkermatter networks in this case are sorted by name, so `network1` comes before `network2`

**Comment by darkermatter** [comment]
So, for ```service2```, eth0 will be ```network1```?

**Comment by aboch** [comment]
@darkermatter Yes.

**Comment by darkermatter** [comment]
Thanks!

It makes sense for rapid development (of docker) but now means we have to think of things like 
```
networks:
  aaa_external_routable_me_first_definitely_me_first: ...
  backend: ...
```

**Comment by aboch** [comment]
@darkermatter Actually what I said is not correct.
The default gateway will be the `network1`'s gateway address.
Which network `eth0` connects the container to depends on the order with which the daemon will connect the networks to the container. So it may be `eth0` or `eth1` and it is independent on the default gateway choosing logic.

**Comment by darkermatter** [comment]
For all intents and purposes, the gateway is the important factor. I just used eth0 as an example as it's typically the interface used for external routable traffic.

If there are various external networks and they are attached to by the daemon independently of how things are specified in docker-compose.yml then the default gateway won't necessarily be ```aaa_external_routable_me_first_definitely_me_first``` due to the order of the 0.0.0.0 entry in the routing table. If that's the case then the documentation is misleading at best.

WRT #21741, from my understanding, I don't think this should be closed as I think there is a deeper solution required than simply changing the documentation.

**Comment by aboch** [comment]
@darkermatter 
No, the default gateway will be via `aaa_external...` network.
Regardless of the order with which the container gets connected to the multiple networks, Docker takes care of programming the default gw in your container based on the lexical order logic, at any connect time. Whenever the container connects to `aaa_external...` network, the default gw for the container is re-programmed via that network.

**Comment by darkermatter** [comment]
Thanks ever so much @aboch  - that's really useful information.

Cheers!

**Pull request merged**